struct.md

```diff
@@ -0,0 +1,297 @@
+But vss-tools does not need to resolve or verify that type reference is correct in VSS 4.0.
+
+```
+
+DeliveryList:
+  datatype: Vehicle.Some.Branch.DeliveryInfo
+  type: sensor
+  description: List of deliveries
+```
+
+## Expectations on VSS implementations (e.g. VISS, KUKSA.val)
+
+It is expected of implementation (long-term) to support atomic read/write/subscribe of signals defined by struct.
+They may support read of parts of signal, e.g. `DeliveryList.Receiver`
+
+## Array Support
+
+It shall be possible to specify that there shall be a struct of the array
+
+
+```
+DeliveryInfo:
+  type: struct
+  description: A struct type containing info for each delivery
+  
+DeliveryInfo.Address:
+  datatype: string
+  type: item
+  description: Destination address
+
+DeliveryInfo.Receiver:
+  datatype: string
+  type: item
+  description: Name of receiver
+
+DeliveryList:
+  datatype: DeliveryInfo[]
+  type: sensor
+  description: List of deliveries
+```
+
+
+### Expectations on VSS implementations (e.g. VISS, KUKSA.val)
+
+For array types (like above) VSS implementations may support several mechanisms
+
+* It is expected that they can support read/write/subscribe of the whole array, i.e. write all or read all in the same request
+* They may optionally support additional operations like
+    * Writing/Reading a single instance, e.g. `DeliveryList[2]` (index mechanism is implementation dependent)
+    * Appending/Deleting individual instances
+    * Searching for instances with specific conditions.
+    
+## Structure in Structure
+
+It shall be possible to refer to a structure type from within a structure
+
+```
+OpenHours:
+  type: struct
+  description: A struct type containing information on open hours
+  
+OpenHours.Open:
+  datatype: uint8
+  type: item
+  max: 24
+  description: Time the address opens
+  
+OpenHours.Close:
+  datatype: uint8
+  type: item
+  max: 24
+  description: Time the address close
+
+DeliveryInfo:
+  type: struct
+  description: A struct type containing info for each delivery
+  
+DeliveryInfo.Address:
+  datatype: string
+  type: item
+  description: Destination address
+
+DeliveryInfo.Receiver:
+  datatype: string
+  type: item
+  description: Name of receiver
+  
+DeliveryInfo.Open:
+  datatype: OpenHours
+  type: item
+  description: When is receiver available
+
+DeliveryList:
+  datatype: DeliveryInfo
+  type: sensor
+  description: List of deliveries
+```
+
+TBD: Where shall the inner struct be defined? Shall it be allowed to define it within a struct as well, or does it need to be defined within a branch like above? If allowed to be defined within a struct, how do we want name resolution to work, only support exact (current) scope and absolute path, or a more flexible setup searching upwards.
+
+I.e. shall the following alternative style (where the struct `OpenHours` is defined within `DeliveryInfo`) be allowed or even preferred?
```

Comment:
This is a much more advanced modeling use case (like inner classes in programming languages which is not that often used). We could extend the specification once the basic modeling of structs is launched and we have feedback on gaps/use cases that need to be covered by future extensions.

Reply:
I have no problem with that limitation. Better to start small and extend as needed. If no-one has any other objection I will remove the "inner struct" example and replace it with a statement similar to:

"It shall not be allowed to define a struct within a struct"